src/umap/Uffd.cpp

```diff
@@ -0,0 +1,288 @@
+
+      if ( !(pollfd[0].revents & POLLIN) )
+        continue;
+
+      int readres = read(m_uffd_fd, &m_events[0], m_max_fault_events * sizeof(struct uffd_msg));
+
+      if (readres == -1) {
+        if (errno == EAGAIN)
+          continue;
+
+        UMAP_ERROR("read failed: " << strerror(errno));
+      }
+
+      assert("Invalid read result returned" && (readres % sizeof(struct uffd_msg) == 0));
+
+      int msgs = readres / sizeof(struct uffd_msg);
+
+      assert("invalid message size" && msgs >= 1 && msgs < m_max_fault_events);
+
+      //
+      // Since uffd page events arrive on the system page boundary which could
+      // be different from umap's page size, the page address for the incoming
+      // events are adjusted to the beginning of the umap page address.  The
+      // events are then sorted in page base address / operation type order and
+      // are processed only once while duplicates are skipped.
+      //
+      for (int i = 0; i < msgs; ++i)
+        m_events[i].arg.pagefault.address &= ~(m_page_size-1);
+
+      std::sort(&m_events[0], &m_events[msgs], less_than_key());
+
+      char* last_addr = nullptr;
+      for (int i = 0; i < msgs; ++i) {
+        if ((char*)(m_events[i].arg.pagefault.address) == last_addr)
+          continue;
+
+        last_addr = (char*)(m_events[i].arg.pagefault.address);
+
+#ifndef UMAP_RO_MODE
+        bool iswrite = (m_events[i].arg.pagefault.flags & (UFFD_PAGEFAULT_FLAG_WP | UFFD_PAGEFAULT_FLAG_WRITE) != 0);
+#else
+        bool iswrite = false;
+#endif
+
+        //
+        // TODO: Since the addresses are sorted, we could optimize the
+        // search to continue from where it last found something.
+        //
+        auto rd = m_rm->containing_region(last_addr);
+        m_buffer->process_page_event(last_addr, iswrite, rd);
+      }
+    }
+    UMAP_LOG(Debug, "Good bye");
+  }
+
+  void Uffd::ThreadEntry() {
+    uffd_handler();
+  }
+
+  Uffd::Uffd( void )
+    :   WorkerPool("Uffd Manager", 1)
+      , m_rm(RegionManager::getInstance())
+      , m_max_fault_events(m_rm->get_max_fault_events())
+      , m_page_size(m_rm->get_umap_page_size())
+      , m_buffer(m_rm->get_buffer_h())
+  {
+    UMAP_LOG(Debug, "\n maximum fault events: " << m_max_fault_events
+                    << "\n            page size: " << m_page_size);
+
+    if ((m_uffd_fd = syscall(__NR_userfaultfd, O_CLOEXEC | O_NONBLOCK)) < 0)
+      UMAP_ERROR("userfaultfd syscall not available in this kernel: "
+          << strerror(errno));
+
+    if (pipe2(m_pipe, 0) < 0)
+      UMAP_ERROR("userfaultfd pipe failed: " << strerror(errno));
+
+    check_uffd_compatibility();
+    m_events.resize(m_max_fault_events);
+
+    start_thread_pool();
+  }
+
+  Uffd::~Uffd() {
+    char bye[5] = "bye";
+
+    write(m_pipe[1], bye, 3);
+
+    stop_thread_pool();
+  }
+
+  void  Uffd::enable_write_protect(
+            void*
+#ifndef UMAP_RO_MODE
+            page_address
+#endif
+        )
```

Comment:
The way I'm reading this, it will be void * followed by a right paren if it is read only mode. Is that what is intended?

Reply:
Correct.  This is one way (perhaps not the best) to prevent compilation warnings of unused arguments when both type and argument are specified when the argument is not used.  Specifying just the type maintains the signature, but no longer produces warnings when the arguments are ifdefed out in the implementation.